lib/AST/ASTImporter.cpp

```diff
@@ -2180,7 +2180,28 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
         if (!ToDescribed)
           return nullptr;
         D2CXX->setDescribedClassTemplate(ToDescribed);
-
+        if (!DCXX->isInjectedClassName() && D->isCompleteDefinition()) {
+          // In a record describing a template the type should be a
+          // InjectedClassNameType. Here the previously set type is updated
+          // to the correct type.
+          // FIXME: The previous type is cleared but not removed from
+          // ASTContext's internal storage, is this problem?
+          D2CXX->setTypeForDecl(nullptr);
+          Importer.getToContext().getInjectedClassNameType(D2CXX,
+              ToDescribed->getInjectedClassNameSpecialization());
+          // Find the injected class name itself and update the type too.
+          CXXRecordDecl *Injected = [&] {
+            for (NamedDecl *Found : D2CXX->noload_lookup(Name)) {
+              CXXRecordDecl *Record = llvm::dyn_cast<CXXRecordDecl>(Found);
```

Comment:
I think you do not need the `llvm::` qualifier and also using auto would be great to avoid type repetition.

Reply:
The `DelayTypeCreation` is already used when the "injected class name" record is created (line 2162). But the correct type is not available at that place (`ToDescribed->getInjectedClassNameSpecialization()` needs to be called). The type in the new record is set temporarily to the "bad" value and updated then to the correct value. If no type is set at all (remove line 2163) it causes problems when importing other things.